src/metabase/driver.clj

```diff
@@ -966,6 +966,29 @@
   dispatch-on-uninitialized-driver
   :hierarchy #'hierarchy)
 
+(defmulti incorporate-auth-provider-details
+  "A multimethod for driver specific behavior required to incorporate response of an auth-provider into the DB details.
+   In most cases this means setting the :password and/or :username based on the auth-provider and its response."
+  {:added "0.32.0" :arglists '([driver auth-provider auth-provider-response details])}
+  dispatch-on-initialized-driver
+  :hierarchy #'hierarchy)
+
+(defmethod incorporate-auth-provider-details :default
+  [_driver _auth-provider _auth-provider-response details]
+  details)
+
+(defmethod incorporate-auth-provider-details :sql-jdbc
+  [_driver auth-provider auth-provider-response details]
+  (case auth-provider
+    (:oauth :azure-managed-identity)
+    (let [{:keys [access_token expires_in]} auth-provider-response]
+      (cond-> (assoc details :password access_token)
+        expires_in (assoc :password-expiry-timestamp (+ (System/currentTimeMillis)
+                                                        ;; renew the password a minute before expiry
+                                                        (* (- (parse-long expires_in) 60) 1000)))))
```

Comment:
Ideally, this `60` should be the same as `metabase.db.data-source/azure-auth-token-renew-slack-seconds`. I'm open to suggestion where this constant should live.

Reply:
As that is something common for `driver` and `data-source` modules, why not to put it into `auth-provider`? I believe there's defined functionality common for the auth providers (`fetch-auth` defmulti), but also provider specific functionality (`fetch-auth :azure-managed-identity`).